### PR TITLE
Chrome trace timeline with sliders.

### PR DESCRIPTION
**Pull request by alanarenda**
This makes progress on #607.

### CODE REVIEW

File: python/ray/WebUI.ipynb [hunk 1]
```diff
@@ -66,20 +199,76 @@
...
+    "        disabled=False,\n",
+    "    )\n",
+    "    \n",
+    "    start_box, end_box, range_slider, time_opt = get_sliders(False)\n",
+    "    display(breakdown_opt)\n",
+    "    display(path_input)\n",
+    "\n",
+    "    def find_trace2html():\n",
+    "        trace2html = os.path.join(os.path.expanduser(\"~\"), \"catapult\", \"tracing\", \"bin\", \"trace2html\")\n",
+    "        assert os.path.exists(trace2html), \"Could not find catapult, please clone it into your home directory from https://github.com/catapult-project/catapult/tree/master/tracing\"\n",
```
Comment (robertnishihara) [inline python/ray/WebUI.ipynb]: This is going to be problematic. The right thing to do is probably to install it when we build Ray. However, for now we can just clone it right here when this function runs (somewhere under `/tmp/`) if it isn't already there.

File: python/ray/experimental/state.py [hunk 2]
```diff
@@ -569,3 +638,27 @@ def workers(self):
                 "stdout_file": worker_info[b"stdout_file"].decode("ascii")
             }
         return workers_data
+
+    def job_length(self):
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: When I call this function without running any tasks, it returns `(9223372036854775807, 0, 0)`. That probably isn't what we want.

Comment (robertnishihara) [inline python/ray/experimental/state.py]: Let's name this `_job_length` so that it isn't exposed to the user.

Also, Please add a docstring explaining precisely what this is measuring.

Also, note that there is a value in Redis already `'redis_start_time'` which is set to the time when the cluster was initiated. Not sure if we want to use this, but it's there (note that it's different from what is currently being used which is presumably the time the first task started).

File: python/ray/experimental/state.py [hunk 3]
```diff
@@ -51,32 +51,32 @@
 class GlobalState(object):
     """A class used to interface with the Ray control state.
 
-    Attributes:
-      redis_client: The redis client used to query the redis server.
-    """
+  Attributes:
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: Undo all of the changes to this file that aren't important for this PR.

File: python/ray/WebUI.ipynb [hunk 4]
```diff
@@ -1,61 +1,194 @@
...
-   "metadata": {},
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
    "source": [
-    "**Evaluate the box below to search for tasks.**"
+    "earliest, latest, num = ray.global_state.job_length()\n",
```
Comment (robertnishihara) [inline python/ray/WebUI.ipynb]: I don't think this box adds very much, I'd consider removing it.

File: python/ray/experimental/state.py [hunk 5]
```diff
@@ -352,69 +358,105 @@ def log_files(self):
 
         return ip_filename_file
 
-    def task_profiles(self, start=None, end=None, num=None):
+    def task_profiles(self, start=None, end=None, num_slice=None, fwd=True):
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: Can you explain what the `fwd` argument is/does? Also document it in the doc string.

File: python/ray/experimental/state.py [hunk 6]
```diff
@@ -352,69 +358,105 @@ def log_files(self):
...
-        heapq.heapify(heap)
-        heap_size = 0
+        # Don't maintain the heap if we're not slicing some number
+        if num_slice is not None:
+            heap = []
+            heapq.heapify(heap)
+            heap_size = 0
+
+        # Set up a param dict to pass the redis command
+        params = {'withscores': True}
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: Let's just use `"` and not `'` in this file for now

File: python/ray/experimental/state.py [hunk 7]
```diff
@@ -352,69 +358,105 @@ def log_files(self):
...
-                                        task_id))
-                        heap_size += 1
                     if event[1] == "ray:get_task" and event[2] == 2:
                         task_info[task_id]["get_task_end"] = event[0]
-                    if (event[1] == "ray:import_remote_function" and
-                            event[2] == 1):
-                        task_info[task_id]["import_remote_start"] = event[0]
-                    if (event[1] == "ray:import_remote_function" and
-                            event[2] == 2):
+                    if event[1] == "ray:import_remote_function" and \
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: Let's not use `\` for multiline things. Instead, this should be formatted something like

```python
if (event[1] == "ray:import_remote_function"
        and event[2] == 1):
```

File: python/ray/WebUI.ipynb [hunk 8]
```diff
@@ -1,61 +1,194 @@
...
+    "    )\n",
+    "   \n",
+    "    INIT_EVENT = 'INIT'\n",
+    "    out_recursion = _EventRecursionContextManager()\n",
+    "   \n",
+    "    def update_wrapper(event):\n",
+    "        if not out_recursion.should_recurse:\n",
+    "            return\n",
+    "       \n",
+    "        with out_recursion:\n",
```
Comment (robertnishihara) [inline python/ray/WebUI.ipynb]: Just out of curiosity, what does "recursion" mean in this context?

File: python/ray/WebUI.ipynb [hunk 9]
```diff
@@ -66,20 +199,76 @@
...
+    "        subprocess.Popen([\"open\", \"-a\", \"Google Chrome\", tmp2])\n",
     "\n",
-    "task_search.on_submit(handle_submit)"
+    "    path_input.on_click(handle_submit)\n",
+    "\n",
+    "task_timeline()"
    ]
+  },
+  {
+   "cell_type": "code",
```
Comment (robertnishihara) [inline python/ray/WebUI.ipynb]: Let's get rid of this empty box at the end.

File: python/ray/experimental/state.py [hunk 10]
```diff
@@ -435,27 +477,34 @@ def task_profiles(self, start=None, end=None, num=None):
...
+      going to chrome://tracing in the chrome web browser and loading the
+      appropriate file.
+
+      Args:
+        path: The filepath to dump the profiling information to.
+        task_info: relevant task profile information for desired time window
+        breakdowns: boolean indicating whether to show tasks in components
+      """
+
+        # TO DO - convert info to deltas
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: Does this TODO still need to be done?

If so, but it's going to be done in a different PR, then this should be formatted along the lines of `# TODO(alanamarzoev):`

**Review by robertnishihara — COMMENTED** [review]
Nice job with this, I left a bunch of comments!

At a high level, the timeline looks amazing, and it'll be really great to get this merged.

Given the verbosity of the code needed to produce the timeline in the UI, we should prioritize hiding the code.

File: python/ray/WebUI.ipynb (hunk 4 above)
Comment (alanamarzoev) [inline python/ray/WebUI.ipynb]: My intended use case was to give a clear understanding of what the possible time windows are for when a user would use the start and end time window boxes.

File: python/ray/WebUI.ipynb [hunk 11]
```diff
@@ -1,61 +1,194 @@
...
    "outputs": [],
    "source": [
-    "object_search = widgets.Text(\n",
-    "    value=\"\",\n",
-    "    placeholder=\"Object ID\",\n",
-    "    description=\"Search for an object:\",\n",
-    "    disabled=False\n",
-    ")\n",
-    "display(object_search)\n",
+    "class _EventRecursionContextManager(object):\n",
```
Comment (alanamarzoev) [inline python/ray/WebUI.ipynb]: There's a single instance of that recursion manager class, out_recursion, whose purpose it is to store whether or not someone is currently in the callback method, as a crude but probably correct way or figuring out whether the next call to the method would be from a human or from code. The recursion manager class implements two methods, __enter__ and __exit__, that allow it to be used as a context manager, so when you have a block annotated with "with out_recursion", when it enters that block it calls out_recursion.__enter__(), which sets some "currently executing" flag to true, and any time you exit, whether through a return or an exception or anything, it calls the out_recursion.__exit__() method, which sets that "currently executing" flag to false. So effectively when a person changes the slider, nothing is currently in the callback, so the slider callback starts to execute, which means setting the flag to true, then the slider sets the value on the text boxes, which tries to execute their callbacks but notices the flag is set to the true so they immediate exit, then the slider callback finishes doing whatever and returns, which sets the flag back to false for when the person next changes anything.

**Comment by robertnishihara** [comment]
@alanamarzoev @ericl, what's the best way to use this when Ray is running on a cluster. Presumably I want to open the web UI in a browser locally on my laptop and then use port forwarding or something to view the jupyter notebook that is running on the cluster. However, how will I view the timeline?

**Comment by ericl** [comment]
Hm, I guess you could also use the programmatic API to dump a file and scp
it back?

**Comment by robertnishihara** [comment]
This PR introduced this bug #756.
